### ISSUE TITLE
Gas limit

### ISSUE THREAD

**Issue by codesmirnov** [BANKEX/web3swift#139]
Hi! 

I'm having problems with transactions, because my gas limit value (4700000) replacing value from gasEstimate (38284). I suppose this happens in the method smartMergeGasLimit.

In this case, I get a successful transaction with warnings:
"Although one or more errors occured [Out of gas] contract execution completed"

How to make it so my value of gas limit not be replaced?

**Comment by shamatar** [comment]
Hello @codesmirnov 

What is your workflow to send a transaction? If you get a gasEstimate, than you can do 
```
transactionIntermediate.send(password: "", options: (..., gasLimit = estimatedGas, ))
```
The whole purpose of smartMerge is to allow overriding by parameters supplied on the last step.

Sincerely, Alex

**Comment by codesmirnov** [comment]
Hi Alex!

I'm using the version 0.8.0

Trying to send like this:

`
var options = Web3Options()
options.to  = address
options.from  = walletAddress
options.gasLimit = BigUInt(SmartContractsManager.gasLimit)
options.gasPrice = gasPrice
options.value = value

method = contract.method(method, parameters: params, options: options)

let result = method.send() // also try method.send(options: options) `

**Comment by shamatar** [comment]
If your do .send() and your limit is higher than estimated gas, then your gas limit will be replaced by gas estimate internally. If you do with .send(options: options) that all the parameters are should be overwritten. What is an expected behavior for your case?

________________________________

**Comment by evgen-povt** [comment]
@shamatar The question is why do you treat an empty gasLimit and the explicitly set gasLimit (when > gasEstimage) in the same way?
If I pass an emty gasLimit it's okay to use gasEstimage. But if I pass gasLimit explicitly and its value is enough for a tx you should not replace it.

`
            if originalOptions != nil, originalOptions!.gasLimit != nil, originalOptions!.gasLimit! < gasEstimage { // original gas estimate was less than what's required, so we check extra options
                if extraOptions != nil, extraOptions!.gasLimit != nil, extraOptions!.gasLimit! >= gasEstimage {
                    return extraOptions!.gasLimit!
                } else {
                    return nil
                }
            } else {
                return gasEstimage
            }
`

**Comment by shamatar** [comment]
@evgen-povt 

I was trying to follow a web3js behavior for similarity, with some "global" default gas limit on the level of web3 object (and it's derivatives like web3.eth). Looks like a rework is needed, will try to fit it into 0.8.1 or in the worst case to 0.9.

Sincerely, Alex

**Comment by shamatar** [comment]
Fixed in 0.9.0

**Comment by GabCas** [comment]
Hi @shamatar,
I have a similar issue after updating to 0.9.0 (I was using 0.8.1 and thought that would fix it).  
I estimate the gas of my method call (46765) and then assign in options `options.gasLimit = gasEstimate + 5` but doesn't seem to overwrite it in the transaction.  I end up getting 'out of gas' error even when trying to put very large gasLimit and gasPrice values.

![screen shot 2018-07-10 at 8 53 30 am](https://user-images.githubusercontent.com/2333673/42511032-c8e7b03a-841e-11e8-8730-6fd67746ee71.png)

Any thoughts on how I can complete the transaction without the 'out of gas' error?

Thanks!

**Comment by shamatar** [comment]
If you want to update a transaction at the last moment before sending, then pass options in the send method. Options are captured by value, so changing it and passing again will do nothing. Would you check this options and if it still breaks I’ll try to add some extra tests

________________________________

**Comment by GabCas** [comment]
Hi @shamatar, I am passing the options in the send method like this:
```
let transactionIntermediate = contract?.method("mymethod", parameters: parameters, options: options)
let gasEstimateResult = transactionIntermediate?.estimateGas(options: options)
guard case .success(let gasEstimate)? = gasEstimateResult else {return}
print(gasEstimate)
options.gasLimit = gasEstimate + 5
let result = transactionIntermediate!.send(password: "mypassword", options: options, onBlock: "latest" )
```

is this what you mean?  I also did estimateGas with options:nil and get the same result.  Lastly, I tried creating a new options variable with the desired values as I didn't quite understand what you mean by "Options are captured by value, so changing it and passing again will do nothing".

Thanks,

**Comment by GabCas** [comment]
This is how I called 'assemble', had to make it public in order to do so:
`let result = transactionIntermediate!.assemble(options: options, onBlock: "latest" ) ` 
`// .send(password: "password", options: options, onBlock: "latest" )`

This is the output:
![screen shot 2018-07-10 at 10 39 23 am](https://user-images.githubusercontent.com/2333673/42517405-a01939a8-842d-11e8-924e-89fc09da2d4d.png)

But I don't see the parameters there... I have another method that uses the same parameters (sha, v, r, s) that works as expected.  However this one always throws the 'out of gas' error.

**Comment by shamatar** [comment]
Is the gas limit equal to what you set?

________________________________

**Comment by GabCas** [comment]
No.  That's the estimated gas.  I'm trying to add 5 more with` options.gasLimit = gasEstimate + 5` so it should be 46770 but it stays at 46765.

**Comment by abusse** [comment]
Seeing the same issue here. No matter what I provide as options, it always defaults to the `gasEstimate` coming from the `smartMergeGasLimit` function.

**Comment by shamatar** [comment]
Please check the fix in 1.0.1 version (either in Pods or on master now)

**Comment by GabCas** [comment]
@shamatar that fixed the issue.

Thanks!
